tobac/segmentation.py

```diff
@@ -455,14 +537,553 @@ def segmentation_timestep(
+            # segmentation approach until the line which states
+            # "#transform segmentation_mask_4 data back to original mask after PBC first-pass ("segmentation_mask_3")"
+            # It's just performed on the buddy box and its data rather than our full domain
+
+            # Set level at which to create "Seed" for each feature in the case of 3D watershedding:
+            # If none, use all levels (later reduced to the ones fulfilling the theshold conditions)
+            if level is None:
+                level = slice(None)
+
+            # transform max_distance in metres to distance in pixels:
+            if max_distance is not None:
+                max_distance_pixel = np.ceil(max_distance / dxy)
+                # note - this doesn't consider vertical distance in pixels
+
+            # mask data outside region above/below threshold and invert data if tracking maxima:
+            if target == "maximum":
+                unmasked_buddies = buddy_data > threshold
+                buddy_segmentation = -1 * buddy_data
+            elif target == "minimum":
+                unmasked_buddies = buddy_data < threshold
+                buddy_segmentation = buddy_data
+            else:
+                raise ValueError("unknown type of target")
+
+            # set markers at the positions of the features:
+            buddy_markers = np.zeros(unmasked_buddies.shape).astype(np.int32)
+            # Buddy boxes are always without PBCs
+            buddy_markers = add_markers(
+                buddy_features,
+                buddy_markers,
+                seed_3D_flag,
+                seed_3D_size,
+                level,
+                PBC_flag="none",
+            )
+
+            # set markers in cells not fulfilling threshold condition to zero:
+            print(np.unique(buddy_markers))
+            buddy_markers[~unmasked_buddies] = 0
+
+            marker_vals = np.unique(buddy_markers)
+
+            # Turn into np arrays (not necessary for markers) as dask arrays don't yet seem to work for watershedding algorithm
+            buddy_segmentation = np.array(buddy_segmentation)
+            unmasked_buddies = np.array(unmasked_buddies)
+
+            # perform segmentation:
+            if method == "watershed":
+                segmentation_mask_4 = watershed(
+                    np.array(buddy_segmentation),
+                    buddy_markers.astype(np.int32),
+                    mask=unmasked_buddies,
+                )
+
+            else:
+                raise ValueError("unknown method, must be watershed")
+
+            # remove everything from the individual masks that is more than max_distance_pixel away from the markers
+            if max_distance is not None:
+                D = distance_transform_edt((markers == 0).astype(int))
+                segmentation_mask_4[
+                    np.bitwise_and(segmentation_mask_4 > 0, D > max_distance_pixel)
+                ] = 0
```

Comment:
I have added a test for `max_distance` not being `None`, but it doesn't actually work with PBCs. The implementation we had for it was simply wrong, so this ended up being an extremely productive suggestion. Lacking an immediate solution, I've tagged it as not implemented and it will raise a `NotImplementedError` when trying to use with PBCs.